src/s2wasm.h

```diff
@@ -229,29 +229,68 @@ class S2WasmBuilder {
   // gets a constant, which may be a relocation for later.
   // returns whether this is a relocation
   // TODO: Clean up this and the way relocs are created from parsed objects
-  bool getRelocatableConst(uint32_t* target) {
+  LinkerObject::Relocation* getRelocatableConst(uint32_t* target) {
     if (isdigit(*s) || *s == '-') {
       int32_t val = getInt();
       memcpy(target, &val, sizeof(val));
-      return false;
-    } else {
-      // a global constant, we need to fix it up later
-      Name name = getStrToSep();
-      LinkerObject::Relocation::Kind kind = isFunctionName(name) ?
-          LinkerObject::Relocation::kFunction :
-          LinkerObject::Relocation::kData;
-      int offset = 0;
-      if (*s == '+') {
-        s++;
-        offset = getInt();
-      } else if (*s == '-') {
-        s++;
-        offset = -getInt();
-      }
-      linkerObj->addRelocation(kind, target,
-                               fixEmLongjmp(cleanFunction(name)), offset);
-      return true;
+      return nullptr;
+    }
+
+    // a global constant, we need to fix it up later
+    Name name = getStrToSep();
+    LinkerObject::Relocation::Kind kind = isFunctionName(name) ?
+        LinkerObject::Relocation::kFunction :
+        LinkerObject::Relocation::kData;
+    int offset = 0;
+    if (*s == '+') {
+      s++;
+      offset = getInt();
+    } else if (*s == '-') {
+      s++;
+      offset = -getInt();
+    }
+    return new LinkerObject::Relocation(
+        kind, target, fixEmLongjmp(cleanFunction(name)), offset);
+  }
+  Expression* relocationToGetGlobal(LinkerObject::Relocation* relocation) {
+    if (!relocation) {
+      return nullptr;
+    }
+
+    auto name = relocation->symbol;
+    auto g = allocator->alloc<GetGlobal>();
+    g->name = name;
+    g->type = i32;
+
+    // Optimization: store any nonnegative addends in their natural place.
+    // Only do this for positive addends because load/store offsets cannot be
+    // negative.
+    if (relocation->addend >= 0) {
+      *relocation->data = relocation->addend;
```

Comment:
I'm not sure LLVM can generate `i32.const foo+2` anyway. I think this is fine for now because we will end up revisiting it for MC and linking support anyway.